src/tools/rbd/Utils.cc

```diff
@@ -815,6 +815,31 @@ int snap_set(librbd::Image &image, const std::string &snap_name) {
   return 0;
 }
 
+bool calc_sparse_extent(const bufferptr &bp, size_t sparse_size, uint64_t length,
+                       size_t &write_offset, size_t &write_length,
+                       size_t &pos, size_t &offset) {
+  size_t shift;
+  if (pos + sparse_size > length) {
+    shift = length - pos;
+  } else {
+    shift = sparse_size;
+  }
+
+  bufferptr page(bp, offset, shift);
+  offset += shift;
+  pos += shift;
+
+  bool page_is_zero = page.is_zero();
+  if (!page_is_zero) {
+    write_length += shift;
+  }
+
+  if ((page_is_zero && write_length != 0) || pos == length) {
```

Comment:
No，when data ---> hole, we need a write.   when reaching the end of the bl, we need a write any way. Please rethink it.